### ISSUE TITLE
Blank screen after syncing Bibles

### ISSUE THREAD

**Issue by alerque** [tobias-klein/ezra-project#24]
I got through with packaging (see #23) and realize there is a new problem. I'm pretty sure it's unrelated to packaging because I'm seeing the same issue on a bone-stock source build (`git clone`, change to working directory, `npm run build-linux`, change to release directory, `./ezra-project`) so this might just be a reflection of the current git HEAD state, but since I don't know what it's about, here's the report:

On first launch on a system with no user installed SWORD modules, I get the normal interface and can hit the settings button for adding repositories, selecting Bibles, etc. Selecting a few and continuing works great, it downloads them, installs them, syncs them, then ... ... nothing. Blank white screen. No console errors, no nothing. The SWORD modules I selected do make it to the user's module directory because they show up in Xiphos.

On second launch I just get the blank white screen. Nothing else.

If I install some more SWORD modules from another app (tested with Xiphos), then re-launch Ezra Project, I get a full interface again with the background greyed out and a modal dialogue box showing that the new modules are being synchronized. That process appears to go fine, but when it's done the entire window turns white again and that's it.

**Comment by tobias-klein** [comment]
Thanks for reporting this issue!

If there's a "blank white screen" that sounds like a crash in a native module, in all likelihood node-sword-interface / SWORD. I have just tested this scenario on my dev box (Ubuntu 18.04 LTS) with the latest master, the "new user context" and a set of randomly selected modules, but I could not reproduce it. I have seen the "blank white screen" issue during development, though and those were always node-sword-interface / native code issues.

Do you still know which exact modules you were installing? Could you provide me a list?

**Comment by alerque** [comment]
> Do you still know which exact modules you were installing? Could you provide me a list?

This is why `package-lock.json` should be committed to the repository for apps like this (not necessarily for libraries, although many people argue it can help).

I'll have to get back to this...

**Comment by tobias-klein** [comment]
I didn't mean node packages but rather which SWORD modules were you selecting for installation in Ezra Project (that made it crash)?

**Comment by alerque** [comment]
Oh, sorry I misunderstood. I don't think it matters at all, I've tried on three different systems and randomly checked off a lot or a few SWORD modules and always get the same result. Once the package download/sync is complete (or if there are existing modules and no changes to sync) it goes straight to blank.

**Comment by tobias-klein** [comment]
Are all systems Arch Linux and is this now with a system-installed package of Sword?
I may need to get your exact environment to be able to reproduce, debug and fix.

**Comment by alerque** [comment]
@alerque Yes all are arch, but no they are not system installed SWORD. When I ran into this issue using my special system packaged version I figured I'd compare it to the "normal" way you are building and tried building from a clean checkout using all the defaults you have for Linux (so your forked node-prune, your git clone and private build of SWORD, whatever node modules `npm` asked for, etc).

```
$ git clone ...
$ cd ezra-project
$ npm run build-linux
$ cd release/ezra-project-linux-x86_64
$ ./ezra-project
```

**Comment by alerque** [comment]
So how do you even start debugging something like this when the app is blank and the console is blank?

**Comment by tobias-klein** [comment]
It's difficult! I'm analyzing which API calls would likely be the culprits and debugging it in the C++ domain of node-sword-interface.

**Comment by tobias-klein** [comment]
I'm actually already getting some issue with the start up.

First of all I get this message:
```
[osboxes@osboxes ezra-project-linux-x64]$ ./ezra-project 
[3290:0307/105334.425250:FATAL:setuid_sandbox_host.cc(157)] The SUID sandbox helper binary was found, but is not configured correctly. Rather than run without sandboxing I'm aborting now. You need to make sure that /home/osboxes/ezra-project/release/ezra-project-linux-x64/chrome-sandbox is owned by root and has mode 4755.
```

Then after putting --no-sandbox in the npm start instruction to at least get a bit further there's an issue with sqlite (visible in Dev Console):

```
Using default database dir /home/osboxes/.config/ezra-project for database access!
/home/osboxes/ezra-p…ix/log-prefix.js:17 [2020-03-07T15:54:38.789Z] Initializing i18n ...
/home/osboxes/ezra-p…ix/log-prefix.js:17 [2020-03-07T15:54:39.161Z] Initializing database ...
/home/osboxes/ezra-p…ix/log-prefix.js:17 [2020-03-07T15:54:39.162Z] Database not yet existing in user directory! Setting up empty database from template.
/home/osboxes/ezra-p…ction-manager.js:81 Uncaught (in promise) Error: Please install sqlite3 package manually
    at ConnectionManager._loadDialectModule (/home/osboxes/ezra-p…ction-manager.js:81)
    at new ConnectionManager (/home/osboxes/ezra-p…ction-manager.js:22)
    at new SqliteDialect (/home/osboxes/ezra-p…/sqlite/index.js:14)
    at new Sequelize (/home/osboxes/ezra-p…ib/sequelize.js:324)
    at DbHelper.getSequelize (/home/osboxes/ezra-p…ers/db_helper.js:81)
    at DbHelper.migrateDatabase (/home/osboxes/ezra-p…ers/db_helper.js:86)
    at DbHelper.initDatabase (/home/osboxes/ezra-p…ers/db_helper.js:36)
    at initDatabase (ezra_init.js:228)
    at initApplication (ezra_init.js:299)
```

This is on Arch Linux 2019.05 downloaded as VM from [here](https://www.osboxes.org/arch-linux/#archlinux-201905-info).

I have actually never seen the first problem before. And the second I saw also on other distributions, but it should technically be gone after `npm run rebuild`.

So, Arch hasn't been treating me well so far ... :-/

**Comment by alerque** [comment]
2019.05 is a pretty old image. Did you update the system with `pacman -Syu` before doing anything else? Always a good idea to start with fresh packages.

I know there is a long history of issues with Electron, sandboxing, and virtualized container conflicts, but for as much as I use containers usually Electron apps are not something I try to run.

**Comment by tobias-klein** [comment]
I now switched to using an Arch Linux Docker image - which easily lets me use the latest Arch distribution. There I'm having another issue now, though. A electron dependency on libffmpeg cannot be resolved:

`./ezra-project: error while loading shared libraries: libffmpeg.so: cannot open shared object file: No such file or directory`

Any pointers?
I did install the "electron" package, which should give me all the dependencies, right?

**Comment by tobias-klein** [comment]
I managed to get Ezra Project on Arch Linux working with the system-supplied Electron package.
The application is starting up fine on a fresh user account, I can install bible translations from the SWORD repos (tried with 4 German and 4 English translations), the install process finishes and everything works as expected. Also after closing and starting again I have new issues, so I'm still not able to reproduce this issue on Arch Linux (latest snapshot 20200306 from Docker).

**Comment by tobias-klein** [comment]
As I'm not able to reproduce then the only way I see to debug this is to add a logging mechanism to node-sword-interface. I'm pretty sure the error has it's rootcause in the native/C++ part, because plain JavaScript won't make an Electron application crash like that.

**Comment by tobias-klein** [comment]
Don't know exactly what was wrong in my Docker container before, but now I'm also able to work with the regular electron version downloaded in the context of the build (Executed via ezra-project binary). So these exact steps work for me on Arch Linux:

```
$ git clone ...
$ cd ezra-project
$ npm run build-linux
$ cd release/ezra-project-linux-x86_64
$ ./ezra-project
```

**Comment by tobias-klein** [comment]
@alerque
Any additional hints on how to reproduce this?

**Comment by alerque** [comment]
@tobias-klein I'm sorry but over the last week in my house the most reproducible thing seems to be poopy diapers; I haven't had a chance to review this. Please do assign the issue to me to come up with a containerized reproduction of some sort.

**Comment by tobias-klein** [comment]
Congrats on the root cause of the diapers, sounds like a more important thing than this :wink:.
Sure I'll assign it to you!

**Comment by tobias-klein** [comment]
Btw. I've been testing with a Docker container based on [this Dockerfile](https://github.com/tobias-klein/ezra-project/blob/master/docker/ArchLinux_Dockerfile).

**Comment by alerque** [comment]
Never mind Docker, I am now (fortunately? unfortunately?) unable to replicate this on any of the systems it was failing on before. Something between 10 days ago and 0.12.0 changed this and my builds work fine.

The Arch packaging is updated to 0.12.0 (currently under Electron 8.1.1, system sword, system node-sword-interface, and more) and works great.

I'm going to close this pending any recurrence or any way I can replicate the problem.
